feat(api): implement from_json method for datahub_actions.event.event_envelope.EventEnvelope

Masterchen09 opened the pull request:
For some local testings I have implemented the from_json method of the EventEnvelope. As the method is using the event registry for getting the event class the EventEnvelope had to be moved from the event.py into a separate file (the event_registry.py imports the event.py, which would then have imported the event_registry.py, which would not have worked 😵).

jjoyce0510 commented on datahub-actions/src/datahub_actions/event/event_registry.py:
```diff
@@ -40,7 +40,7 @@ def from_class(cls, clazz: MetadataChangeLogClass) -> "MetadataChangeLogEvent":
     def from_json(cls, json_str: str) -> "Event":
         json_obj = json.loads(json_str)
-        return cls.from_obj(json_obj, True)
+        return cls.from_obj(json_obj)
```
Why remove the True here?

jjoyce0510 commented on datahub-actions/src/datahub_actions/event/event_envelope.py:
```diff
@@ -0,0 +1,60 @@
+    # Convert a json event envelope back into the object.
+    @classmethod
+    def from_json(cls, json_str: str) -> "EventEnvelope":
+        json_obj = json.loads(json_str)
```
Looks great- thanks for implementing!

jjoyce0510 commented on datahub-actions/src/datahub_actions/event/event_registry.py:
```diff
@@ -60,7 +60,7 @@ def from_class(cls, clazz: EntityChangeEventClass) -> "EntityChangeEvent":
     def from_json(cls, json_str: str) -> "EntityChangeEvent":
         json_obj = json.loads(json_str)
-        return cls.from_obj(json_obj, True)
+        return cls.from_obj(json_obj)
```
Do you see the 'parameters' field being correctly deserialized from json back into the EntityChangeEvent?

jjoyce0510 reviewed (approved):
Overall this LGTM

Masterchen09 commented:
I had some issues with the value field of the GenericAspect in the MetadataChangeLogEvent, therefore I have removed the tuple parameter. The value field of the GenericAspect is a bytes field, but the JSON representation of the field is a string. After some debugging I have figured out that the avrogen package cannot handle string-encoded bytes field (correctly) when the tuples parameter is true (the path I followed was the following:
[from_obj](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/dict_wrapper.py#L39) calls [with_tuple_union](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/avrojson.py#L31) -> self.fastavro is True when tuples is true -> from_obj also calls [from_json_object](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/avrojson.py#L92) which calls [_generic_from_json](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/avrojson.py#L233) -> _generic_from_json calls [_union_from_json](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/avrojson.py#L312) which calls [validates](https://github.com/acryldata/avro_gen/blob/f2c65f4c15ea6bcbc535c627ddeb3f340be9bd06/avrogen/avrojson.py#L85) -> validates accepts strings for bytes fields only if self.fastavro is false).

I am not sure if this might be an issue of the avrogen package, but it worked with the removed tuples parameter, therefore I was happy at that point. 😁

For the EntityChangeEvent we might not need to change the tuples parameter, because there should not be any bytes field in there (or at least at the moment?).

I have also added some tests for the from_json method of the two event classes and the event envelope, I think these are helpful to verify that the methods are working correctly.

The parameter deserialization now also works for the EnitityChangeEvent and is using the hack from the Kafka source (not nice, but it works...?). I have also noticed that the from_obj method of the events does not return an Event instance, instead an instance of the parent class is returned, therefore the from_class method is used to construct a valid Event instance (originally I only wanted to quickly test something...therefore I did not noticed that the instance was not an instance of the Event).

To have the hack with the parameters only at one place I have also updated the Kafka event source to use the from_json method of the EntityChangeEvent (the Kafka event source no longer needs it's own logic for the deserialization of the message).

jjoyce0510 commented:
Amazing! Looking over this again now

jjoyce0510 commented on datahub-actions/src/datahub_actions/event/event_registry.py:
```diff
@@ -60,7 +60,19 @@ def from_class(cls, clazz: EntityChangeEventClass) -> "EntityChangeEvent":
+        if parameters:
+            event._inner_dict["__parameters_json"] = parameters
+
+        return event
```
Thanks for consolidating this!

jjoyce0510 commented on datahub-actions/src/datahub_actions/plugin/source/kafka/kafka_event_source.py:
```diff
@@ -75,29 +69,9 @@ def build_metadata_change_log_event(msg: Any) -> MetadataChangeLogEvent:
-    if "parameters" in json_payload:
-        event._inner_dict["__parameters_json"] = json_payload["parameters"]
-    return event
+    return EntityChangeEvent.from_json(payload.get("value"))
```
Nice!!

jjoyce0510 commented on datahub-actions/src/datahub_actions/event/event_registry.py:
```diff
@@ -60,7 +60,19 @@ def from_class(cls, clazz: EntityChangeEventClass) -> "EntityChangeEvent":
+        return event
 
     def as_json(self) -> str:
         json_obj = self.to_obj()
```
I guess the downside to keeping this in __parameters_json is that it's not easy to access for the consumer :(

jjoyce0510 replied:
Without first serializing to json of course

jjoyce0510 commented on datahub-actions/tests/unit/event/test_event_registry.py:
```diff
@@ -0,0 +1,84 @@
+    event_json_str = event.as_json()
+    event_from_json = EntityChangeEvent.from_json(event_json_str)
+
+    diff = deepdiff.DeepDiff(event, event_from_json, ignore_order=True)
```
Amazing. Thank you so much for adding these!

jjoyce0510 commented on datahub-actions/tests/unit/event/test_event_envelope.py:
```diff
@@ -0,0 +1,61 @@
+import json
+import pprint
+
+import deepdiff
```
Thank you!
